support webrtc-datachannel sdp exchange

duiniuluantanqin opened the pull request:
Currently, SRS does not support WebRTC data channel. If a client requests to establish a data channel connection, it will cause an error in the browser.
```
Failed to set remote answer sdp: The order of m-lines in answer doesn't match order in offer. Reject
```

The SDP response from SRS before modification is as follows:
```
v=0
...
a=group:BUNDLE 0 1
m=audio 9 UDP/TLS/RTP/SAVPF 111
...
m=video 9 UDP/TLS/RTP/SAVPF 106 114
...
```

After modification:
```
v=0
...
a=group:BUNDLE 0 1 2
m=audio 9 UDP/TLS/RTP/SAVPF 111
...
m=video 9 UDP/TLS/RTP/SAVPF 106 114
...
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
...
```

**Doubtful:**
~~Is datachannel considered as a media stream? Does it count as a track? If not, it should not be placed inside the `SrsMediaDesc` class, but should be separated.~~

---------

`TRANS_BY_GPT3`

duiniuluantanqin committed:
support webrtc-datachannel sdp exchange

duiniuluantanqin committed:
Update srs_app_rtc_api.cpp

duiniuluantanqin committed:
Update srs_app_rtc_conn.cpp

johzzy commented on trunk/src/app/srs_app_rtc_source.cpp:
```diff
@@ -2302,6 +2308,10 @@ SrsRtcTrackDescription* SrsRtcSourceDescription::find_track_description_by_ssrc(
         return audio_track_desc_;
     }
 
+    if (application_track_desc_ && application_track_desc_->has_ssrc(ssrc)) {
```
Why is "has ssrc" used here?

`TRANS_BY_GPT3`

duiniuluantanqin replied:
Well, the datachannel does not have the ssrc attribute, so it should be removed.

`TRANS_BY_GPT3`

duiniuluantanqin committed:
Update srs_app_rtc_source.cpp

chundonglinlin committed:
HttpConn: judge nb_chunk no memory address. (#3465)

Co-authored-by: john <hondaxiao@tencent.com>

winlinvip committed:
Support WHIP and WHEP player. v5.0.147 and v6.0.35 (#3460)

Co-authored-by: chundonglinlin <chundonglinlin@163.com>
Co-authored-by: panda <542638787@qq.com>

winlinvip committed:
WHIP: Support DELETE resource for Larix Broadcaster. v5.0.148 v6.0.36 (#3427)

* WHIP: Support DELETE resource.
* Support push by Larix.
* FLV: Disable stash buffer for realtime.
* WHEP: Fix muted issue.

-------

Co-authored-by: chundonglinlin <chundonglinlin@163.com>
Co-authored-by: panda <542638787@qq.com>

winlinvip committed:
Fix release bug. Release v6.0.36

duiniuluantanqin committed:
support webrtc-datachannel sdp exchange

xiaozhihong committed:
RTC: Call on_play before create session, for it might be freed for timeout. v5.0.149, v6.0.37  (#3455)

Co-authored-by: winlin <winlin@vip.126.com>
Co-authored-by: ChenGH <chengh_math@126.com>

yashwardhan-jyani committed:
Remove unneccessary NULL check in srs_freep. v5.0.150, v6.0.38 (#3477)

Co-authored-by: winlin <winlin@vip.126.com>
Co-authored-by: john <hondaxiao@tencent.com>

winlinvip committed:
WebRTC: Support configure CANDIDATE by env (#3470)

In dockerfile, we can set the default RTC candidate to env:

```
ENV SRS_RTC_SERVER_CANDIDATE=\$CANDIDATE
CMD ["./objs/srs", "-e"]
```

When starts a docker container, user can setup the candidate by env:

```
docker run --rm -it --env CANDIDATE=1.2.3.4 ossrs/srs:5
```

We should parse the content of SRS_RTC_SERVER_CANDIDATE as env variable name and parse it again.

---------

Co-authored-by: chundonglinlin <chundonglinlin@163.com>
Co-authored-by: pengfei.ma <pengfei.ma@ctechm.com>

chundonglinlin committed:
WebRTC: Error message carries the SDP when failed. v5.0.151, v6.0.39 (#3450)

Co-authored-by: winlin <winlin@vip.126.com>
Co-authored-by: ChenGH <chengh_math@126.com>

duiniuluantanqin committed:
support application for webrtc play

suzp1984 commented on trunk/research/players/js/srs.page.js:
```diff
@@ -116,6 +118,38 @@ function build_default_rtc_url(query) {
+    console.log('?api=x to overwrite WebRTC API(1985).');
+    console.log('?schema=http|https to overwrite WebRTC API protocol.');
+
+    var server = (!query.server)? window.location.hostname:query.server;
```
why not give `?a:b` some spaces, ref line #131 `'http:' ? '1985' : '1990'`. They are so crowded, are they?

winlinvip commented:
Thank you, but the PR has been closed for too long a time.
